Documentation about choosing metric types (#750)
mozilla/glean: docs/user/adding-new-metrics.md

### For how long do you need to collect this data?

Think carefully about how long the metric will be needed, and set the `expires` parameter to disable the metric at the earliest possible time. 
This is an important component of Mozilla's [lean data practices][lean-data].

When the metric passes its expiration date (determined at build time), it will automatically stop collecting data.

When a metric's expiration is within in 14 days, emails will be sent from `telemetry-alerts@mozilla.com` to the `notification_emails` addresses associated with the metric.
At that time, the metric should be removed, which involves removing it from the `metrics.yaml` file and removing uses of it in the source code.
Removing a metric does not affect the availability of data already collected by the pipeline.

If the metric is still needed after its expiration date, it should go back for [another round of data review](https://wiki.mozilla.org/Firefox/Data_Collection) to have its expiration date extended.

### When should Glean automatically clear the measurement?

The `lifetime` parameter of a metric defines when its value will be cleared. There are three lifetime options available:

> - `ping` (default): The metric is cleared each time it is sent in the ping.

Author: Very good point.  Let's use "submit" and not "collect" to be consistent with the APIs on that, though.

> - `ping` (default): The metric is cleared each time it is sent in the ping.

Reviewer: That's even better!

  This is the most common case, and should be used for metrics that are highly dynamic, such as things computed in response to the user's interaction with the application.
  

> - `application`: The metric is related to an application run, and is cleared every time the application restarts.

Reviewer 2: ```suggestion
- `application`: The metric is related to an application run, and is cleared after the application restarts and any Glean-owned ping, due at startup, is submitted.
```

  
![Metric lifetime timeline](metric-lifetime-timeline.svg)

- **A. Ping lifetime, set on change**: The value isn't included in Ping 1, because Glean doesn't know about it yet.  It is included in the first ping after being recorded (Ping 2), which causes it to be cleared.

- **B. Ping lifetime, set on init and change**: The default value is included in Ping 1, and the changed value is included in Ping 2, which causes it to be cleared.  It therefore misses Ping 3, but when the application is started, it is recorded again and it is included in Ping 4.  However, this causes it to be cleared again and it is not in Ping 5.

- **C. Application lifetime, set on change**: The value isn't included in Ping 1, because Glean doesn't know about it yet. After the value is changed, it is included in Pings 2 and 3, but then due to application restart it is cleared, so it is not included until the value is manually toggled again.

- **D. Application, set on init and change**: The default value is included in Ping 1, and the changed value is included in Pings 2 and 3. Even though the application startup causes it to be cleared, it is set again, and all subsequent pings also have the value.

> - **E. User, set on change**: The default value is missing from Ping 1, but since `user` lifetime metrics aren't cleared unless the user profile is reset, it is included in all subsequent pings.

Reviewer 2: ```suggestion
- **E. User, set on change**: The default value is missing from Ping 1, but since `user` lifetime metrics aren't cleared unless the user profile is reset (e.g. on Android, when the product is uninstalled), it is included in all subsequent pings.
```

> - **F. User, set on change**: Since `user` lifetime metrics aren't cleared unless the user profile is reset, it is included in all pings.

Reviewer: ```suggestion
- **F. User, set on init and change**: Since `user` lifetime metrics aren't cleared unless the user profile is reset, it is included in all pings.
```

> - **F. User, set on init and change**: Since `user` lifetime metrics aren't cleared unless the user profile is reset, it is included in all pings.

Reviewer 2: Would you consider adding a note mentioning that `user` lifetime metrics will be kept on sending even if the metric was set in Run 1 but never in Run 2?

Note that for all of the metric configurations, the toggle of the preference off and on during Ping 4 is completely missed.  If you need to create a ping containing one, and only one, value for this metric, consider using a [custom ping](pings/custom.html) to create a ping whose lifetime matches the lifetime of the value.

#### What if none of these lifetimes are appropriate?

If the timing at which the metric is sent in the ping needs to closely match the timing of the metrics value, the best option is to use a [custom ping](pings/custom.html) to manually control when pings are sent.

### What should this new metric be called?

#### Reuse names from other applications

> There's a lot of value using the same name for analogous metrics collected across different products. For example, BigQuery makes it simple to join columns with the same name across multiple tables. Therefore, we encourage you to investigate if a similar metric is already being collected by another product and use the same naming whenever feasible.

Reviewer 2: Thinking again of this, if some metric is really shared across products of the same platform, maybe devs should consider writing a component/library using glean?

Author: I think it's a good idea to add a note about that, but it may not always be feasible.  Think of "open new tab", for example.  That's handled by a completely different code in Desktop vs. Fenix that it probably doesn't make sense to merge, even though they both have that "concept" of "open new tab".

I think maybe we say something like:

"Therefore, we encourage you to investigate if a similar metric is already being collected by another product.  If it is, there may be an opportunity for code reuse across these products.  But even if that's not feasible, use the same metric name for similar actions and concepts whenever possible."

Reviewer 2: Good enough! Thank you for your patience!

If you need to know the time between multiple distinct actions that aren't a simple "begin" and "end" pair, consider using an [event](metrics/event.html).

### Do you need to know the order of events relative to other events?

If you need to know the order of actions relative to other actions, such as, the user performed tasks A, B, and then C, and this is meaningfully different from the user performing tasks A, C and then B, (in other words, the order is meaningful beyond just the *fact* that a set of tasks were performed), use an [event metric](metrics/event.html).

Importantly, events are the most expensive metric type to record, transmit, store and analyze, so they should be used sparingly, and only when none of the other metric types are sufficient for answering your question.

### For how long do you need to collect this data?

Think carefully about how long the metric will be needed, and set the `expires` parameter to disable the metric at the earliest possible time. 
This is an important component of Mozilla's [lean data practices][lean-data].

> When a metric is to expire in 14 days, emails will be sent from `telemetry-alerts@mozilla.com` to the `notification_emails` addresses associated with the metric.

Reviewer 3: ```suggestion
When a metric's expiry is within 14 days, emails will be sent from `telemetry-alerts@mozilla.com` to the `notification_emails` addresses associated with the metric.
```

  
- `application`: The metric is related to an application run, and is cleared every time the application restarts.
  This should be used for things that are constant during the run of an application, such as the operating system version.
  In practice, these metrics are generally set during application startup.
  A common mistake---using the `ping` lifetime for these type of metrics---means that they will only be included in the first ping sent during a particular run of the application.
  
- `user`: The metric is part of the user's profile.
  This should be used for things that change only when the user's profile is created.
  It is rare to use this lifetime outside of some metrics that are built in to Glean, such as `client_id`.

While lifetimes are important to understand for all metric types, they are particularly important for the metric types that record single values and don't aggregate on the client (`boolean`, `string`, `labeled_string`, `string_list`, `datetime` and `uuid`), since these metrics will send the "last known" value and missing the earlier values could be a form of unintended data loss.

#### A lifetime example

> Let's work through an example to see how these lifetimes play out in practice. Let's suppose we have a user preference, "turbo mode", which defaults to `false`, but the user can turn it to `true` at any time.  We want to know when this flag is `true` so we can measure its affect on other metrics in the same ping.  In the following diagram, we look at a time period that sends 4 pings across two separate runs of the application. We assume here, that like Glean's built-in [metrics ping](pings/metrics.html), the developer writing the metric isn't in control of when the ping is submitted. 

Reviewer 2: Do we want to know _when_ it gets to true or what's its status when a ping is generated? I think the latter, right?

Author: Yes, the latter.  That's what I meant by: "We want to know when this flag is `true` so we can measure its affect on other metrics in the same ping."  Any suggestions to make that more clear?

Reviewer 2: Sorry, I think I don't :\ Let's leave it as it is.

In this diagram, the ping "lifetimes" are represented as rectangles, but the moment the ping is "submitted" is represented by its right edge. The user changes the "turbo mode" setting from `false` to `true` at 3 seconds.
  
```mermaid
gantt
  axisFormat %S
  title Metric lifetimes
  dateFormat  s

  section Metrics<br/>pings

>     Ping 1 :active, p1, 0, 2s

Reviewer 2: Would it make sense to note the value reported by these metrics ping?

Author: Yes, but I'm not sure how to fit that in the visualization -- the value is different for each of the metric examples.  (I'm feeling a bit underwhelmed by mermaid for this -- I might need to fire up Inkscape and do it by hand, but I was trying to avoid that...)

    Ping 2 :active, p2, 2, 3s
    Ping 3 :active, p3, 5, 10s
    Ping 4 :active, p4, 15, 5s
    Ping 5 :active, p5, 20, 5s

  section Process
	Run 1     :done, a1, 0, 10s
	Run 2     :done, a2, 15, 10s

  section Interaction
    ↓ Turns on turbo mode : 3

  section Ping<br/>(change)
	True      :3, 2s

  section Ping<br/>(init and change)
    False     :0, 2s
    True      :3, 2s
    True      :15, 5s

  section Application<br/>(change)
    True      :3, 12s  

  section Application<br/>(init and change)
    False     :0, 3s
    True      :3, 12s
    True      :15, 10s

  section User<br/>(change)
    True      :3, 22s

> ```

Author: ![image](https://user-images.githubusercontent.com/38294/75825259-ff5a8a00-5d72-11ea-9e03-e2a65af50980.png)

Reviewer 2: Can we break out this lifetime example part to a separate PR?

Author: It's a little mixed up, particularly since the lifetime section was moved as part of this change.  Maybe we can do another round of review all together, and if it gets stalled, I'll pick the two apart.

- **Ping (change)**: The lifetime is set to `ping` and the value is recorded into the metric when it changes. The value isn't included in Ping 1, because Glean doesn't know about it yet.  It is included in the first ping after being recorded (Ping 2), which causes it to be cleared and it is not included in any subsequent pings.

> - **Ping (init and change)**: The lifetime is set to `ping` and the value is recorded at application initialization and when it changes. The default value is included in Ping 1, and the changed value is included in Ping 2, which causes it to be cleared.  It therefore misses Ping 3, but when the application is started, it is recorded again and it is included in Ping 4.  However, this causes it to be cleared again and it is not in Ping 5.

Reviewer 2: Is "Ping" here meaning a custom ping? Or the 'metrics' ping? I think we should disambiguate as it gets confusing: 'ping' is a lifetime and a ping name

Author: It's ping lifetime.  Will disambiguate.

- **Application (change)**: The lifetime is set to `application` and the value is recorded into the metric with it changes. The value isn't included in Ping 1, because Glean doesn't know about it yet. After the value is changed, it is included in Pings 2 and 3, but then due to application restart it is cleared, so it is not included in any subsequent pings.

- **Application (init and change)**: The lifetime is set to `application` and the value is recorded at application initialization and when it changes. The default value is included in Ping 1, and the changed value is included in Pings 2 and 3. Even though the application startup causes it to be cleared, it is set again, and all subsequent pings also have the value.

> - **User (change)**: The lifetime is set to `user` and the value is recorded when it changes. Since `user` lifetime metrics aren't cleared unless the user profile is reset, it is included in all subsequent pings.

Reviewer 2: It's worth noting that this fails to capture if the value was flipped between two pings (i.e. flag is on, then off, then again on ). We should probably highlight the disadvantages of this option as well, otherwise we risk this seen as the holy grail of the solutions

Author: Good point -- missing flipping twice during a ping lifetime is a disadvantage of *every* approach, though.  It's definitely worth noting that.

Reviewer 2: Oh, right :)

Reviewer 3: The solution there is to cut a ping when the value changes, but that's a second example where the ping's schedule _is_ under the data owner's control.

When adding a new metric, the workflow is:

* Consider the question you are trying to answer with telemetry, and choose the [metric type](metrics/index.md) and parameters to use.
* Add a new entry to [`metrics.yaml`](#The-metricsyaml-file).
* Add code to your project to record into the metric by calling the Glean SDK.

> **Important**: Any new data collection requires documentation and data-review.
This is also required for any new metric automatically collected by the Glean SDK.

## Choosing a metric type

> ### Is it a simple (scalar) measurement?

Reviewer 2: I think @chutten and DS should really have a look at this too!

If the value is true or false, use a [boolean metric](metrics/boolean.html).

If the value is a string, use a [string metric](metrics/string.html).  If you want to store multiple strings in a group, use a [string list metric](metrics/string_list.html).

> If you need to record a closely-related group of strings, or you don't know the things the strings are related to at build time, use a [labeled string metric](metrics/labled_strings.html).

Author: Honestly, it's hard to find motivating examples for this one.  Even the real-world use cases that do exist probably shouldn't have been using this.

@Dexterp37 : Do you remember the motivation for this one?

> If you need to record a closely-related group of strings, or you don't know the things the strings are related to at build time, use a [labeled string metric](metrics/labled_strings.html).

Reviewer 2: Mh, no, I don't really remember. If in doubt, maybe we can disable that and re-enable if needed. I'm not sure that's currently used. The only example I can find is the [one in our docs](https://mozilla.github.io/glean/book/user/metrics/labeled_strings.html#configuration)

> If you need to record a closely-related group of strings, or you don't know the things the strings are related to at build time, use a [labeled string metric](metrics/labled_strings.html).

Author: There is this usage in sync telemetry in the wild: https://github.com/mozilla-mobile/android-components/blob/f046b61335ad4f2f026e65536421540d31f33b97/components/support/sync-telemetry/metrics.yaml#L116

It's fine -- I don't think it's a problem we should ask to fix, but the use case doesn't seem very compelling, somehow.

> If you need to record a closely-related group of strings, or you don't know the things the strings are related to at build time, use a [labeled string metric](metrics/labled_strings.html).

Reviewer 2: Oh , right! They are using this metric as there's no better way, right now, to report errors for them :( I agree with you: that's not a very compelling use-case.

For all of these simple value types, you should consider how the value changes in the application and how that relates to the ping it is being sent in. When the ping is submitted, it will contain the "last known" value for the metric, and then the value is reset according to its [lifetime](#What-is-the-lifetime-of-the-data).

> If you need to know the current state of a value, such as a user preference, to provide context for other metrics, be aware that the lifetime of the value may not be the same as the lifetime of the ping it is in. You can address this either with a custom ping that matches the lifetime of the value, or by using a user lifetime metric so the value will persist across pings.

Author: Aside from just being hard to explain:

I think for metrics that want to record the status of a preference (like the environment stuff on desktop) so that it can be related to other metrics, I think the best recommendation in our toolbox is a user lifetime metric, but I know others have *opinions* about that.

Reviewer 2: > I think the best recommendation in our toolbox is a user lifetime metric, but I know others have _opinions_ about that.

:-P 

I personally think that the status of a preference (exactly like the environment on desktop) is `application` lifetime: the _specific_ environment on Desktop is built at startup (ok, in a staged way) and lives throughout the life of Firefox (or as long as it is changed).

A relevant example, for desktop environment, is search defaults: it's set [later at startup](https://searchfox.org/mozilla-central/rev/91f6c02fcf4c16f78fdc4417f61f192688294066/toolkit/components/telemetry/app/TelemetryEnvironment.jsm#1418) and, in the rare circumstance of pings being generated before that, it will not be present in pings. Still, that's not a "user lifetime": it lives as long as the application lives.

I think the core of the issue here is understanding what it's being measured and when/how it should be available: that's indeed an hard task :)

Reviewer: Maybe expanding on the gfx metric example of the `compositor` would be a good way to illustrate the dangers of using application lifetime metrics with the metrics ping, and that it requires some extra care? That seemed like a good real-world case that shows where application lifetime metrics can be weird.

Reviewer 3: We can leave it as a warning that the lifetime might not match and let the rest of the documentation suggest solutions appropriate to the reader. Or we could be a bit more precise about "If it is important that every ping contain some value for this metric, it should have a 'ping' lifetime and be sent on a custom ping you control."

### Are you counting things?

If you want to know how many times something happened, use a [counter metric](metrics/counter.html).  If you are counting a group of related things, or you don't know what all of the things to count are at build time, use a [labeled counter metric](metrics/labeled_counter.html).

If you need to know when the things being counted happened relative to other things, use an [event](metrics/event.html).

### Are you measuring time?

If you need to record an absolute time, use a [datetime metric](metrics/datetime.html). Datetimes are recorded in the user's local time, along with a timezone offset from UTC.

> If you need to record how long something takes you have a couple of options:

Reviewer 3: This is the first instance of this structure in the guide. Everything else has been a paragraph of related metric types with increasing complexity, and this is more of a list. I suggest we pick a structure for the whole guide and use it throughout.

Also "a couple" to me means "two". There are three options presented : )

> If you need to measure the total time spent doing a particular task, look to the [timespan metric](metric/timespan.html). Note that it can only be used on a single thread.

Reviewer 3: Might be worth a thumbnail of what happens if you misuse it.

> If you need to measure the relative occurrences of many timings, use a [timing distribution](metric/timing_distribution.html). It builds a histogram of timing measurements, and is safe to use from multiple threads.

Reviewer 3: Does this mean "start" and "stopandaccumulate" can be on different threads? Is a distribution with a single sample a suggested solution for someone who needs a cross-thread timespan?

Author: How about:

```
This metric can safely measure multiple concurrent timespans on different threads.
```

(Cross-thread timespans aren't really the intended use case, though since we use integer ids to identify the timespans, we can't stop you).  But I think this level of detail belongs in the reference of the timing distribution metric itself.

> If you need to know the time between distinct events that aren't a simple "begin" and "end" pair, consider using an [event](metric/event.html).

Reviewer 2: At first I was... "eugh" with this sentence. But it's really a legitimate use-case I didn't think about! Good catch.

Reviewer 3: Any "Eugh" is a candidate for a new metric type to me. Let's say a "timing list" where each step in the multi-part process is represented as a count of nanos from the first `start()`. ...but if what we're hoping is to hear from people who want this metric type, we can't provide a workaround for them in this guide.

Author: I think if you want to answer, how long was spent between performing action A and action B, events are exactly the thing that was designed for.  I think the proposed timing list metric might be redundant to all of that.  Let me reword this to be a little more on-the-nose about the events use case.

### Do you need to know the order of events relative to other events?

> If you need to know the order of events relative to other events, such as, the user did A, then B, then C, use an [event metric](metric/event.html).

Reviewer 2: ```suggestion
If you need to know the order of actions relative to other actions, such as, the user did A, then B, then C, use an [event metric](metric/event.html).
```

What about this to reduce the repetition of "event"?

> If you need to know the order of actions relative to other actions, such as, the user did A, then B, then C, use an [event metric](metric/event.html).

Reviewer 3: I'd add something about how "And this is meaningfully different than if the user did A then C then B." The order has to be meaningful, otherwise you could just count A and B and C.

> Events are the most expensive metric type to record, transmit, store and analyse, so they should be used sparingly, and only when one of the other metric types aren't sufficient for answering your question.

Author: Also perhaps a higher potential for accidental PII, but maybe that's hard to back up with facts.

Reviewer 2: Good point.

Another thing that might be relevant (but that I really struggle to phrase properly) is that events push the computational efforts to the "analysis" step (e.g. think of some of our users using events to compute, at analysis time, the "foreground duration"). Other metric types make sure the effort is at the SDK level, handled by the SDK.

Reviewer 3: It can be backed up with facts: events are structured strings. They contain more bytes per record than any other metric type besides labeled strings, and contain even more bytes overall because a labeled string can only store one string per label: events can record different values each time you hit record.

Put another way, if you were looking for a way to exfiltrate a copy of wikipedia from Fenix to the Data Platform via the Glean SDK, events would be the way you'd choose to do it.

> ### For how long do you need to collect this data?

Reviewer 3: One last segment in this guide should probably exhort the reader to peruse the options in the documentation. This guide will not be complete as niche datatypes do not make sense to have the reader be guided to (like the UUID metric type, for example). And then consider reaching out to us on #glean:mozilla.org, or to request a new Metric Type if all else fails.

> Think carefully about how long the metric will be needed, and set the `expires` parameter to expire the metric at the soonest time possible.

Reviewer 2: Is it worth linking to our [lean data](https://www.mozilla.org/en-US/about/policy/lean-data/) website here, for a more in depth read? Specifically https://www.mozilla.org/en-US/about/policy/lean-data/stay-lean/ (point 4) might be relevant to this particular sentence.

> Think carefully about how long the metric will be needed, and set the `expires` parameter to expire the metric at the soonest time possible.

Reviewer: ```suggestion
Think carefully about how long the metric will be needed, and set the `expires` parameter to expire the metric at the earliest possible time.
```

Reviewer: nit: 'soonest' just sounded weird here for some reason.

> When a metric is about to expire, emails will be sent to the `notification_emails` addresses associated with the metric.

Reviewer 3: I think if we specify from what email address and what service these emails will come we will build trust in them.

- `user`: The metric is part of the user's profile.
  This should be used for things that change only when the user's profile is created.

>   It is rare to use this lifetime outside of some metrics that are built in to Glean, such as `client_id`.

Author: The lifetime language here hasn't changed, but probably needs to be improved as it relates to the other discussion of lifetimes in the "simple value" section (which is where getting the lifetimes matters most, IMHO).

Reviewer 2: Yes, good idea. What improvements did you have in mind? Any relevant example we might just embed in the docs?

Reviewer: As I said above, I think any 'real-life' examples we could provide that illustrate our intentions with lifetimes would be useful.
